File: app/src/main/java/io/smileyjoe/applist/fragment/AppListFragment.java
package io.smileyjoe.applist.fragment;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.Optional;

import io.smileyjoe.applist.R;
import io.smileyjoe.applist.adapter.AppDetailAdapter;
import io.smileyjoe.applist.databinding.FragmentAppListBinding;
import io.smileyjoe.applist.enums.Direction;
import io.smileyjoe.applist.enums.Page;
import io.smileyjoe.applist.object.AppDetail;
import io.smileyjoe.applist.util.Db;
import io.smileyjoe.applist.util.Notify;
import io.smileyjoe.applist.viewholder.AppDetailViewHolder;

public class AppListFragment extends Fragment {

    public interface Listener {
        void onLoadComplete(Page page, int position, int appCount);
    }

    public interface ScrollListener {
        void onScroll(Direction direction);
    }

    public interface ItemSelectedListener extends AppDetailViewHolder.ItemSelectedListener{}

    private static final String EXTRA_PAGE = "page";
    private static final String EXTRA_POSITION = "position";
    private Page mPage;
    private AppDetailAdapter mAppDetailAdapter;
    private boolean mLoading = true;
    private int mPosition;
    private Optional<Listener> mListener;
    private Optional<ScrollListener> mScrollListener;
    private ItemSelectedListener mItemSelectedListener;
    private FragmentAppListBinding mView;

    public AppListFragment() {
    }

    public static AppListFragment newInstance(Page page, int position) {
        AppListFragment fragment = new AppListFragment();
        Bundle args = new Bundle();
        args.putSerializable(EXTRA_PAGE, page);
        args.putInt(EXTRA_POSITION, position);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        mPage = (Page) getArguments().getSerializable(EXTRA_PAGE);
        mPosition = getArguments().getInt(EXTRA_POSITION);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        mView = FragmentAppListBinding.inflate(getLayoutInflater(), container, false);

        setupAdapter();

        mView.recyclerAppDetails.setLayoutManager(new LinearLayoutManager(getContext()));
        mView.recyclerAppDetails.setAdapter(mAppDetailAdapter);
        mView.recyclerAppDetails.setOnScrollChangeListener(this::dispatchScroll);

        populateList();
        handleDisplayView();

        return mView.getRoot();
    }

    private void dispatchScroll(View view, int scrollX, int scrollY, int oldScrollX, int oldScrollY){
        if (scrollY > oldScrollY) {
            mScrollListener.ifPresent(l -> l.onScroll(Direction.DOWN));
        } else if (scrollY < oldScrollY) {
            mScrollListener.ifPresent(l -> l.onScroll(Direction.UP));
        }
    }

    public void setupAdapter() {
        mAppDetailAdapter = new AppDetailAdapter(new ArrayList<AppDetail>(), mPage);
        mAppDetailAdapter.onSave(appDetail -> appDetail.save(getActivity()));
        mAppDetailAdapter.onDelete(appDetail -> appDetail.delete(getActivity()));
        mAppDetailAdapter.onItemSelected(mItemSelectedListener);
    }

    public void setListener(Listener listener) {
        mListener = Optional.of(listener);
    }

    public void setScrollListener(ScrollListener scrollListener) {
        mScrollListener = Optional.of(scrollListener);
    }

    public void onItemSelected(ItemSelectedListener listener){
        mItemSelectedListener = listener;
    }

    private void populateList() {
        Db.getDetailReference(getActivity()).addValueEventListener(new AppDetailsEventListener());
    }

    private void handleDisplayView() {
        if (mLoading) {
            mView.progressLoading.setVisibility(View.VISIBLE);
            mView.recyclerAppDetails.setVisibility(View.GONE);
            mView.textEmpty.setVisibility(View.GONE);
        } else if (mAppDetailAdapter.hasApps()) {
            mView.progressLoading.setVisibility(View.GONE);
            mView.recyclerAppDetails.setVisibility(View.VISIBLE);
            mView.textEmpty.setVisibility(View.GONE);
        } else {
            mView.progressLoading.setVisibility(View.GONE);
            mView.recyclerAppDetails.setVisibility(View.GONE);
            mView.textEmpty.setVisibility(View.VISIBLE);
        }
    }

    private class AppDetailsEventListener implements ValueEventListener {
        @Override
        public void onDataChange(DataSnapshot dataSnapshot) {
            mAppDetailAdapter.update(mPage.getApps(getContext(), dataSnapshot));

            mLoading = false;
            handleDisplayView();

            mListener.ifPresent(listener -> listener.onLoadComplete(mPage, mPosition, mAppDetailAdapter.getItemCount()));
        }

        @Override
        public void onCancelled(DatabaseError databaseError) {
            Notify.error(getActivity(), R.string.error_database_read_failed);
        }
    }
}

File: app/src/main/java/io/smileyjoe/applist/fragment/AppListFragment.kt
package io.smileyjoe.applist.fragment

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import io.smileyjoe.applist.R
import io.smileyjoe.applist.adapter.AppDetailAdapter
import io.smileyjoe.applist.databinding.FragmentAppListBinding
import io.smileyjoe.applist.db.Db
import io.smileyjoe.applist.enums.Page
import io.smileyjoe.applist.extensions.Compat.getSerializableCompat
import io.smileyjoe.applist.util.Notify
import io.smileyjoe.applist.viewholder.AppDetailViewHolder

/**
 * Fragment containing a list of apps, filtered by the [Page] details
 */
class AppListFragment : Fragment() {

    /**
     * Callback for when the page is loaded
     */
    fun interface OnLoadComplete {
        /**
         * The page is loaded
         *
         * @param page the page that loaded
         * @param appCount how many apps are in the list
         */
        fun onLoadComplete(page: Page, appCount: Int)
    }

    /**
     * The user selected an item in the list
     *
     * @see [AppDetailViewHolder.OnItemSelected]
     */
    fun interface OnItemSelected : AppDetailViewHolder.OnItemSelected

    companion object {
        private const val EXTRA_PAGE: String = "page"

        /**
         * Get an instance of the fragment set to the specified [Page]
         *
         * @param page page details to show
         * @return instance of the fragment
         */
        fun newInstance(page: Page): AppListFragment =
            AppListFragment().apply {
                arguments = Bundle().apply {
                    putSerializable(EXTRA_PAGE, page)
                }
            }
    }

    var isLoading: Boolean = true
    lateinit var page: Page
    lateinit var appDetailAdapter: AppDetailAdapter
    lateinit var binding: FragmentAppListBinding
    var onLoadComplete: OnLoadComplete? = null
    var onItemSelected: OnItemSelected? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)

        page = requireArguments().getSerializableCompat(EXTRA_PAGE, Page::class.java)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentAppListBinding.inflate(layoutInflater, container, false)

        setupAdapter()

        binding.recyclerAppDetails.apply {
            layoutManager = LinearLayoutManager(context)
            adapter = appDetailAdapter
        }

        populateList()
        handleDisplayView()

        return binding.root
    }

    /**
     * Setup the adapter
     */
    private fun setupAdapter() {
        appDetailAdapter = AppDetailAdapter(
            page = page,
            saveListener = { app -> app.db.save(requireActivity()) },
            deleteListener = { app -> app.db.delete(requireActivity()) },
            onItemSelected = this@AppListFragment.onItemSelected
        )
    }

    /**
     * Add a listener to the db, this will fire once for the first load, and then will make
     * sure the list is kept up to date, eg, if an installed app is saved, it will immediately
     * be added to the saved list.
     */
    private fun populateList() {
        Db.getDetailReference(requireActivity())?.addValueEventListener(AppDetailsEventListener())
    }

    /**
     * Setup which view to display based on the information we have. Either the list will show,
     * or the loader, or an error message.
     */
    private fun handleDisplayView() {
        // only one shows at a time, so set them all to GONE to start //
        var progressVisibility = View.GONE
        var recyclerVisibility = View.GONE
        var textVisibility = View.GONE

        // set only the needed view to visible //
        if (isLoading) {
            progressVisibility = View.VISIBLE
        } else if (appDetailAdapter.hasApps()) {
            recyclerVisibility = View.VISIBLE
        } else {
            textVisibility = View.VISIBLE
        }

        // update all the visibilities //
        binding.apply {
            progressLoading.visibility = progressVisibility
            recyclerAppDetails.visibility = recyclerVisibility
            textEmpty.visibility = textVisibility
        }
    }

    /**
     * Db listener for any updates to the firebase objects.
     * </p>
     * Updates the adapter on any changes and notifies the calling class
     * that data has loaded
     */
    inner class AppDetailsEventListener : ValueEventListener {
        override fun onDataChange(snapshot: DataSnapshot) {
            appDetailAdapter.items = page.getApps(requireContext(), snapshot)
            isLoading = false
            handleDisplayView()
            onLoadComplete?.onLoadComplete(page, appDetailAdapter.itemCount)
        }

        override fun onCancelled(error: DatabaseError) {
            Notify.error(requireActivity(), R.string.error_database_read_failed)
        }
    }
}